### DISCUSSION TITLE
Recognizing contributors

### DISCUSSION THREAD

**Discussion by abelsiqueira** [JuliaSmoothOptimizers/Organization#41] [Ideas]
The most important thing here is to consider that we want to foster a community, so recognizing contributors is essencial.
There are a few ways to recognize contributors:

- Include them in the CITATION.cff file (which will be used to generate the DOIs in Zenodo);
- Include them in the relevant software paper such as JOSS;
- Add an [allcontributors](https://allcontributors.org/docs/en/overview) section in the repo and/or site;
- Add a section on the website similar to the allcontributors (hard to maintain).

There is some debate about the definition of contributor and authors.
The allcontributors approach recognizes every kind of contribution, especially non-code, which often is ignored but is extremely important for a healthy community.
However, the CITATION.cff and papers usually only have fields for **authors** (CFF might change, check in 6 months).

The author role is sometimes accepted as equal to contributors, and in some cases associated more traditionally with the owners of the packages, regardless of who contributed later.
Notice that this doesn't touch the research paper, which has its own debate related to whether the developers are included in the paper or not.
So the question is who is included in the CITATION.cff and JOSS paper as author? If an external contributor writes a new solver inside JSOSolvers, should they be included as authors?
What about people who don't contribute to the code _per se_, but help with ideas? What about just documentation?

I think we mostly took the conservative approach, with some fault on "update speed". E.g., @tmigot  was not included in the JSOSolvers repo CITATION.bib, but he has the most commits there (being fixed as I write).

**Comment by abelsiqueira** [comment]
Related: https://github.com/SciML/ColPrac

**Reply by tmigot** [reply]
Hi @abelsiqueira, thanks for opening this debate.

I would classify like this:
- **authors**: original team creating the package, and it could be increased if there is a major refactoring, e.g., ADNLPModels.jl. They are authors of the JOSS paper.
- **main contributors**: developers of an important feature for the package. For instance, for JSOSolvers.jl, it would be adding a new solver or reworking the solver organization. An additionaly criterion could be to expect/trust the main contributors to also be a maintainer of the package.
- **contributors**: anyone that has a merged PR, or helped in any way. 

Github is naturally listing people in the commit history. I am not too fond of the allcontributor-thing at the repo level as I suspect it will be additional repetitive work on us, but at the organization level it could be nice.

If I understand the ColPrac recommendation, a "contributor" is someone with merge rights. I am not sure I fully agree with this for JSO as currently the merge rights are attributed at the organization level and not the package level.

I can review the existing packages and make a list of the potential changes to BIB or CFF if we were adding main contributors to see how it would look like.

**Reply by abelsiqueira** [reply]
I like your list, but indeed I would call the middle one "maintainers", and give them merge rights (for that package).
And just to clarify, "maintainers will receive merge rights" and not "maintainers are those with merge rights", otherwise all of us would be maintainers of everything.

Do you have an example of a maintainer that is not an author? I feel like they have done enough to get upgraded to author. I mean, the name author is not great in this context.

The allcontributors thing is pretty limited in that sense. It also involves setting up in a generic enough way for all packages, and it might be too much.

Feel free to check some packages, but take your time. I think this will take some time to be implemented.

**Reply by tmigot** [reply]
I have two examples of maintainers that didn't contribute sufficiently (I think although it can be discussed) to be authors:
- @geoffroyleconte on LDLFactorizations.jl;
- myself on SolverBenchmark.jl or BenchmarkProfiles.jl

The case of an author that is not a maintainers also exist of course.

**Comment by tmigot** [comment]
A closely related topic. A dozen of packages do not have the "and contributors" in the authors list in the bib file. Should this be fixed? Should we do the same with CFF files?

Examples: FluxNLPModels.jl & KnetNLPModels.jl & MUMPS.jl
& NLPModelsJuMP.jl & PartiallySeparableNLPModels.jl & PartitionedStructures.jl & PartitionedVectors.jl & Percival.jl & QRMumps.jl & RipQP.jl.

**Comment by tmigot** [comment]
Some news here, the citations have been revised in all the repos and the authors updated! There aren't so many new potential authors, so we probably can manually maintain this.
